### PR TITLE
Creating versions setting section

### PR DESCRIPTION
**Pull request by john-science**
## Description

Removing metadata section of settings files and creating a "versions" section under settings.  Some of the problems this attempts to solve are:

* #1234 - The ARMI settings "metadata" is hard to access in the settings file (Yamlize doesn't allow it).
* #1234 - Since the metadata is nearly impossible to access in the DB as well.
* #1267 - Let's just ditch the metadata
* Making it easier (more obvious) how ARMI-based Applications and Plugins could record their versions during a run.

**NOTE** - This is SORT-OF an API-breaking change.  So it comes with an ARMI version bump.  However, I expect that it won't actually break anything downstream. It will _allow_ applications and plugins to record their versions when writing a settings YAML file.  But... they don't have to or anything.

---

## Checklist

- [X] This PR has only one purpose or idea.
- [X] Tests have been added/updated to verify that the new/changed code works.

- [X] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [X] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

- [X] The [release notes](https://terrapower.github.io/armi/release/index.html) (location `doc/release/0.X.rst`) are up-to-date with any bug fixes or new features.
- [X] The documentation is still up-to-date in the `doc` folder.
- [X] The dependencies are still up-to-date in `setup.py`.

### CODE REVIEW

**Comment by john-science** [comment]
@drewj-usnctech @albeanth @sombrereau  Just FYI.

**Comment by john-science** [comment]
@opotowsky  Even if everyone likes this PR, I will want to talk to you about WHEN will be a good time to merge it.  (I think it wouldn't affect downstream applications, but still we'll want to touch base.)

File: armi/settings/fwSettings/globalSettings.py [hunk 1]
```diff
@@ -33,87 +33,88 @@
...
 CONF_TARGET_K = "targetK"  # lots of things use this
+CONF_TIGHT_COUPLING = "tightCoupling"
+CONF_TIGHT_COUPLING_MAX_ITERS = "tightCouplingMaxNumIters"
+CONF_TIGHT_COUPLING_SETTINGS = "tightCouplingSettings"
+CONF_TRACE = "trace"
 CONF_TRACK_ASSEMS = "trackAssems"
+CONF_UNIFORM_MESH_MINIMUM_SIZE = "uniformMeshMinimumSize"
+CONF_USER_PLUGINS = "userPlugins"
 CONF_VERBOSITY = "verbosity"
+CONF_VERSIONS = "versions"
```
Comment (albeanth) [inline armi/settings/fwSettings/globalSettings.py:116]: Love the alphabetizing here.

**Comment by albeanth** [comment]
I like this PR but I think it will have (very) minor implications on downstream applications. I'll touch base offline.

File: armi/settings/fwSettings/globalSettings.py (hunk 1 above)
Comment (john-science) [reply armi/settings/fwSettings/globalSettings.py:116]: Yeah, I couldn't help myself.  I didn't _have_ to do the cleanup. But there were ~50 constants all in no perceptible order.  So hard to read and understand.

File: armi/settings/settingsIO.py [hunk 2]
```diff
@@ -311,7 +323,13 @@ def _preprocessYaml(self, settingData):
             else:
                 cleanedData[settingObj.name] = settingDatum
 
+        # add ARMI version to the settings YAML
+        if CONF_VERSIONS not in cleanedData:
+            cleanedData[CONF_VERSIONS] = {}
+        cleanedData[CONF_VERSIONS]["armi"] = version
```
Comment (opotowsky) [inline armi/settings/settingsIO.py:329]: Where does `version` come from here? Am I being a little dense?

File: armi/settings/fwSettings/globalSettings.py [hunk 3]
```diff
@@ -33,87 +33,88 @@
 
 
 # Framework settings
```
[the remark this answers is not in the document]
Comment (john-science) [reply armi/settings/fwSettings/globalSettings.py:35]: Yeah, I just alphabetized. 

There were dozens of them, and it was hard to figure out if the one I wanted to add already existed in some form or another.

File: armi/settings/settingsIO.py (hunk 2 above)
Comment (john-science) [reply armi/settings/settingsIO.py:329]: At the top of the file it says:

```python
from armi.meta import __version__ as version
```

This is where we store the ARMI version in `armi/meta.py`.

Comment (john-science) [reply armi/settings/settingsIO.py:329]: I could remove the rename.  Perhaps it would have been more clear without it?

Comment (opotowsky) [reply armi/settings/settingsIO.py:329]: no no I like it! I was just assuming it was something that would show up in the diff. This makes sense.

**Review by opotowsky — APPROVED** [review]

**Review by albeanth — APPROVED** [review]

**Pull request merged**